lnode/syncrandaddr.go

```diff
@@ -0,0 +1,299 @@
+			if err != nil {
+				log.Warningf("requestAddrNonce to neighbor %v error: %v\n", nbr.GetAddr(), err)
+				return
+			}
+
+			var addr160 common.Uint160
+			addr160.SetBytes(addrNonceReply.Address)
+
+			// here we get next expected nonce = last nonce + 1
+			myNonce, err := localNode.TxnPool.GetNonceByTxnPool(addr160)
+			if err != nil {
+				log.Info("localNode.TxnPool.GetNonceByTxnPool err: ", err)
+			} else {
+				myNonce = myNonce - 1 // deduct 1 to get current last nonce
+			}
+
+			if myNonce < addrNonceReply.Nonce {
+				addrTxnRelpy, err := localNode.requestSyncAddrTxn(nbr, addrNonceReply.Address, myNonce)
+				if err != nil {
+					log.Info("localNode.requestSyncAddrTxn err: ", err)
+				} else {
+					// update txn pool
+					for _, pbTxn := range addrTxnRelpy.Transactions {
+
+						txn := &transaction.Transaction{Transaction: pbTxn}
+
+						localNode.TxnPool.AppendTxnPool(txn)
+					}
+				}
+
+			}
+
+		}(neighbor)
+	}
+
+}
+
+// send neighbor a REQ_ADDR_NONCE request message, and wait for reply.
+func (localNode *LocalNode) requestAddrNonce(remoteNode *node.RemoteNode) (*pb.AddrNonce, error) {
+
+	msg, err := NewReqAddrNonceMessage()
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := localNode.SerializeMessage(msg, false)
+	if err != nil {
+		return nil, err
+	}
+
+	replyBytes, err := remoteNode.SendBytesSyncWithTimeout(buf, syncReplyTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	replyMsg := &pb.AddrNonce{}
+	err = proto.Unmarshal(replyBytes, replyMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return replyMsg, nil
+}
+
+// send to a neighbor a REQ_SYNC_ADDR_TXN request message, and wait for reply with timeout
+// return the reply messge RPL_SYNC_ADDR_TXN
+func (localNode *LocalNode) requestSyncAddrTxn(remoteNode *node.RemoteNode, addr []byte, nonce uint64) (*pb.Transactions, error) {
+
+	msg, err := NewReqSyncAddrTxnMessage(addr, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := localNode.SerializeMessage(msg, false)
+	if err != nil {
+		return nil, err
+	}
+
+	replyBytes, err := remoteNode.SendBytesSyncWithTimeout(buf, syncReplyTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	replyMsg := &pb.Transactions{}
+	err = proto.Unmarshal(replyBytes, replyMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("Sync txn by random address, get ", len(replyMsg.Transactions), "txns from neighbor")
+
+	return replyMsg, nil
+}
+
+// build a REQ_ADDR_NONCE request message
+func NewReqAddrNonceMessage() (*pb.UnsignedMessage, error) {
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_REQ_ADDR_NONCE,
+	}
+
+	return msg, nil
+
+}
+
+// build RPL_ADDR_NONCE reply message
+// reply with rand address my txn pool's with this address' nonce.
+func NewRplAddrNonceMessage(tp *pool.TxnPool) (*pb.UnsignedMessage, error) {
+	addrList := tp.GetAddressList()
+	addrCount := len(addrList)
+	if addrCount <= 0 {
+		return nil, fmt.Errorf("no transaction in my pool")
```

Comment:
Node should return a normal response when txpool is empty.

Reply:
yes. fixed.